### PR TITLE
devicestate: implement checkFDEFeatures()

### PR DESCRIPTION
**Pull request by mvo5**
Run the fde-setup hook with "op":"features" as part of the install
to ensure that we test that the hardware actually supports
encryption. Later this features operation may also alter the way
that the install happens.

### CODE REVIEW

**Review by anonymouse64 — APPROVED** [review]
lgtm thanks for this

File: overlord/devicestate/devicemgr.go [hunk 1]
```diff
@@ -1453,6 +1441,39 @@ func (m *DeviceManager) runFDESetupHook(op string, params *boot.FDESetupHookPara
...
+	// returns any {"features":[...]} reply we consider the
+	// hardware supported. If the hook errors or if it returns
+	// {"error":"hardware-unsupported"} we don't.
+	output, err := m.runFDESetupHook("features", &boot.FDESetupHookParams{})
+	if err != nil {
+		return err
+	}
+
+	// first check for an error
+	var errRes struct {
```
Comment (pedronis) [inline overlord/devicestate/devicemgr.go:1455]: we can parse things into one struct with both error and features fielads

File: overlord/devicestate/devicemgr.go [hunk 2]
```diff
@@ -1453,6 +1441,39 @@ func (m *DeviceManager) runFDESetupHook(op string, params *boot.FDESetupHookPara
...
+		Error string `json:"error"`
+	}
+	if err := json.Unmarshal(output, &errRes); err == nil && errRes.Error != "" {
+		return fmt.Errorf("cannot use hook, it returned error: %v", errRes.Error)
+	}
+	// then check  for features, any feature reply is considered good
+	var featuresRes struct {
+		Features []string `json:"features"`
+	}
+	if err := json.Unmarshal(output, &featuresRes); err == nil {
```
Comment (pedronis) [inline overlord/devicestate/devicemgr.go:1465]: should we accept  {}  as valid and corresponding to {"features": []}?  I don't mind too much either way, atm we do but the tests don't check

File: overlord/devicestate/devicestate_install_mode_test.go [hunk 3]
```diff
@@ -940,7 +961,58 @@ func (s *deviceMgrInstallModeSuite) TestInstallCheckEncryptedErrors(c *C) {
...
+	for _, tc := range []struct {
+		hookOutput  string
+		expectedErr string
+	}{
+		// invalid json
+		{"xxx", `cannot parse hook output: "xxx"`},
+		// no output is invalid
+		{"", `cannot parse hook output: ""`},
+		// specific error
+		{`{"error":"failed"}`, `cannot use hook, it returned error: failed`},
```
Comment (pedronis) [inline overlord/devicestate/devicestate_install_mode_test.go:994]: we need to check for {"features": 1} for example

File: overlord/devicestate/devicestate_install_mode_test.go [hunk 4]
```diff
@@ -940,7 +961,58 @@ func (s *deviceMgrInstallModeSuite) TestInstallCheckEncryptedErrors(c *C) {
...
+		// valid
```
Comment (pedronis) [inline overlord/devicestate/devicestate_install_mode_test.go:995]: as I said we need to decide about {}

**Review by pedronis — COMMENTED** [review]
some comments

File: overlord/devicestate/devicemgr.go [hunk 5]
```diff
@@ -1453,6 +1441,39 @@ func (m *DeviceManager) runFDESetupHook(op string, params *boot.FDESetupHookPara
...
 }
 
+func (m *DeviceManager) checkFDEFeatures(st *state.State) error {
+	// Run fde-setup hook with "op":"features". If the hook
+	// returns any {"features":[...]} reply we consider the
+	// hardware supported. If the hook errors or if it returns
+	// {"error":"hardware-unsupported"} we don't.
+	output, err := m.runFDESetupHook("features", &boot.FDESetupHookParams{})
+	if err != nil {
+		return err
```
Comment (pedronis) [inline overlord/devicestate/devicemgr.go:1451]: this error is logged already right?

File: overlord/devicestate/devicemgr.go [hunk 6]
```diff
@@ -1453,6 +1441,39 @@ func (m *DeviceManager) runFDESetupHook(op string, params *boot.FDESetupHookPara
...
+	}
+
+	// first check for an error
+	var errRes struct {
+		Error string `json:"error"`
+	}
+	if err := json.Unmarshal(output, &errRes); err == nil && errRes.Error != "" {
+		return fmt.Errorf("cannot use hook, it returned error: %v", errRes.Error)
```
Comment (pedronis) [inline overlord/devicestate/devicemgr.go:1459]: should we log this?

**Review by pedronis — COMMENTED** [review]
logging questions

File: overlord/devicestate/devicemgr.go (hunk 2 above)
Comment (mvo5) [reply overlord/devicestate/devicemgr.go:1465]: I think we should be strict initially, updated the code

**Review by pedronis — APPROVED** [review]
thx, see my questions about logging

File: overlord/devicestate/devicemgr.go (hunk 5 above)
Comment (mvo5) [reply overlord/devicestate/devicemgr.go:1451]: This error will be eventually bubble up into the handler in doSetupRunSystem() and appear as a task error. Do you think we should further log this using a logger?

Comment (pedronis) [reply overlord/devicestate/devicemgr.go:1451]: I think it's eaten in checkEncryption actually? that's why I asked about this to help debugging

Comment (mvo5) [reply overlord/devicestate/devicemgr.go:1451]: I misread the code, the error was discarded but I pushed a commit that reads it and logs it via a logger.Noticef()

Comment (mvo5) [reply overlord/devicestate/devicemgr.go:1451]: Code is updated, I merge this now but happy to do a followup if you want to see tweaks to the message or the way this is done

Comment (anonymouse64) [reply overlord/devicestate/devicemgr.go:1451]: @mvo5 where was the code updated ? I don't see any logger.Noticef in this PR ...

**Pull request merged**